src/mono/mono/metadata/loader-internals.h

```diff
@@ -62,6 +72,54 @@ struct _MonoAssemblyLoadContext {
 };
 #endif /* ENABLE_NETCORE */
 
+struct _MonoMemoryManager {
+	MonoDomain *domain;
+	// Whether the MemoryManager can be unloaded on netcore; should only be set at creation
+	gboolean collectible;
+	// Whether this is a singleton or generic MemoryManager
+	gboolean is_generic;
+	// Whether the MemoryManager is in the process of being freed
+	gboolean freeing;
+
+	// Entries moved over from the domain:
+
+	// If taking this with the loader lock, always take this second
+	// Currently unused, we take the domain lock instead
+	MonoCoopMutex lock;
+
+	MonoMemPool *mp;
+	MonoCodeManager *code_mp;
+
+	GPtrArray *class_vtable_array;
+
+	// !!! REGISTERED AS GC ROOTS !!!
+	// Hashtables for Reflection handles
+	MonoGHashTable *type_hash;
+	MonoConcGHashTable *refobject_hash;
+	// Maps class -> type initializaiton exception object
+	MonoGHashTable *type_init_exception_hash;
+	// Maps delegate trampoline addr -> delegate object
+	//MonoGHashTable *delegate_hash_table;
+	// End of GC roots
+};
+
+struct _MonoSingletonMemoryManager {
+	MonoMemoryManager memory_manager;
+
+	// Parent ALC, NULL on framework
+	MonoAssemblyLoadContext *alc;
+};
+
+#ifdef ENABLE_NETCORE
+struct _MonoGenericMemoryManager {
```

Comment:
nit: Maybe add a comment here that explains the design - _eg_ "If a generic type from one ALC is instantiated with types from another ALC, a memory manager that points to both ALCs will own the memory for the instantiation, similar to MonoImageSet"

Reply:
I don't actually do anything with this type here, but I'm happy to add that comment in the follow-up pr where it actually gets used.

Reply:
Also, it might be better to keep these types internal to memory-manager.c, so the rest of the code doesn't need the casts.